src/core/document.js

```diff
@@ -821,103 +821,95 @@ class PDFDocument {
     });
   }
 
-  get xfaDatasets() {
+  get _xfaStreams() {
     const acroForm = this.catalog.acroForm;
     if (!acroForm) {
-      return shadow(this, "xfaDatasets", null);
+      return null;
     }
 
     const xfa = acroForm.get("XFA");
+    const entries = {
+      "xdp:xdp": null,
+      template: null,
+      datasets: null,
+      config: null,
+      connectionSet: null,
+      localeSet: null,
+      stylesheet: null,
+      "/xdp:xdp": null,
+    };
     if (xfa instanceof BaseStream && !xfa.isEmpty) {
-      try {
-        const xdp = stringToUTF8String(xfa.getString());
-        return shadow(this, "xfaDatasets", new DatasetReader({ xdp }));
-      } catch (_) {
-        warn("XFA - Invalid utf-8 string.");
-        return shadow(this, "xfaDatasets", null);
-      }
+      entries["xdp:xdp"] = xfa;
+      return entries;
     }
 
     if (!Array.isArray(xfa) || xfa.length === 0) {
       return null;
     }
 
     for (let i = 0, ii = xfa.length; i < ii; i += 2) {
-      if (xfa[i] !== "datasets") {
+      let name;
+      if (i === 0) {
+        name = "xdp:xdp";
+      } else if (i === ii - 2) {
+        name = "/xdp:xdp";
+      } else {
+        name = xfa[i];
+      }
+
+      if (!entries.hasOwnProperty(name)) {
         continue;
       }
       const data = this.xref.fetchIfRef(xfa[i + 1]);
       if (!(data instanceof BaseStream) || data.isEmpty) {
         continue;
       }
-      try {
-        const datasets = stringToUTF8String(data.getString());
-        return shadow(this, "xfaDatasets", new DatasetReader({ datasets }));
-      } catch (_) {
-        warn("XFA - Invalid utf-8 string.");
-        return shadow(this, "xfaDatasets", null);
-      }
+      entries[name] = data;
     }
-
-    return shadow(this, "xfaDatasets", null);
+    return entries;
   }
 
-  get xfaData() {
-    const acroForm = this.catalog.acroForm;
-    if (!acroForm) {
-      return null;
+  get xfaDatasets() {
+    const streams = this._xfaStreams;
+    if (!streams) {
+      return shadow(this, "xfaDatasets", null);
     }
-
-    const xfa = acroForm.get("XFA");
-    const entries = {
-      "xdp:xdp": "",
-      template: "",
-      datasets: "",
-      config: "",
-      connectionSet: "",
-      localeSet: "",
-      stylesheet: "",
-      "/xdp:xdp": "",
-    };
-    if (xfa instanceof BaseStream && !xfa.isEmpty) {
+    for (const key of ["datasets", "xdp:xdp"]) {
+      const stream = streams[key];
+      if (!stream) {
+        continue;
+      }
       try {
-        entries["xdp:xdp"] = stringToUTF8String(xfa.getString());
-        return entries;
+        const str = stringToUTF8String(stream.getString());
+        const data = Object.fromEntries([[key, str]]);
```

Comment:
Nit: Can't you create this object directly, using the `{ [key]: str }` notation, or does that perhaps not work correctly with the development viewer (because of the SystemJS dependency)?